numpy/core/src/multiarray/mapping.c

```diff
@@ -624,6 +624,7 @@ prepare_index(PyArrayObject *self, PyObject *index,
                 for (i=0; i < n; i++) {
                     Py_DECREF(nonzero_result[i]);
                 }
+                Py_DECREF(arr);
                 goto failed_building_indices;
```

Comment:
Incidentally, are there any race concerns with putting a `Py_DECREF` after the `PyErr_SetString`? I don't use the C API much, but I vaguely recall admonishments that any time you decref it can allow Python to be re-entered via finalizers. I put it here for now to be next to the other decref.

Reply:
I think this should be fine, finalizers stash exceptions that are currently in flight while they are running.